### PR TITLE
Adds managed schemas

### PR DESCRIPTION
**Pull request by alxarch**
### Background

We will be tracking Panther managed schemas in the panther-analysis repo
This commit includes an initial export of Panther native schemas to YAML and the `Makefile` rules to build a release artifact for managed schemas

### Changes

* Adds `schemas/` directory to hold managed schema YAML definitions.
* Adds `managed-schemas.zip` target to `Makefile` that produces a managed schemas archive to be included in releases

### Testing

* Built schemas artifact

### CODE REVIEW

File: schemas/logs/aws/cloud_trail.yml
Comment (kostaspap) [inline schemas/logs/aws/cloud_trail.yml:75]: `null.String`?

File: schemas/logs/aws/guard_duty.yml
Comment (kostaspap) [inline schemas/logs/aws/guard_duty.yml:70]: Shouldn't be in here.

File: schemas/logs/aws/guard_duty.yml
Comment (kostaspap) [inline schemas/logs/aws/guard_duty.yml:4]: GuardDuty parser extracts some indicators e.g. AWS ARNs however this file doesn't declare any. Will the table in the datalake have these fields?

File: schemas/logs/aws/cloud_trail.yml
Comment (alxarch) [reply schemas/logs/aws/cloud_trail.yml:75]: It's an auto-generated field description as generated by the `mage build:schemas` command. We can discuss how to handle auto-descriptions and make a PR on `panther` repo, re-export and merge here.

File: schemas/logs/aws/guard_duty.yml
Comment (alxarch) [reply schemas/logs/aws/guard_duty.yml:70]: See above

File: schemas/logs/aws/guard_duty.yml
Comment (alxarch) [reply schemas/logs/aws/guard_duty.yml:4]: The `native` parser config will force panther to resolve the schema from the `logtypes.Entry`. So yes, schemas in the data lake will be the same as they were. As we progress porting more 'native' log types to  either direct YAML-based defintions or non-legacy formats, the schema definitions here will be getting 'richer'.

File: schemas/logs/aws/cloud_trail.yml
Comment (alxarch) [reply schemas/logs/aws/cloud_trail.yml:75]: If the field has no description it uses `{FIELD_NAME} field {FIELD_TYPE}`

Comment (kostaspap) [reply schemas/logs/aws/cloud_trail.yml:75]: > If the field has no description it uses `{FIELD_NAME} field {FIELD_TYPE}`

I think it could be just `{FIELD_NAME} field`. The type of it is anyway already present in the field schema, and we can avoid awkward descriptions like this one.

**Review by lindsey-w — COMMENTED** [review]
The `Makefile` updates and the `schemas` organization LGTM

**Review by wey-chiang — COMMENTED** [review]
I've left some general comments of things that seemed wrong to me.

Since I am not well-read on schemas, deferring approval to another reviewer.

File: schemas/logs/aws/tests/cloud_trail_digest_tests.yml
Comment (nhakmiller) [inline schemas/logs/aws/tests/cloud_trail_digest_tests.yml:1]: Two concerns on tests:

1. The test is in a different file/location than the log schema. Is the reason to split these out just to make concatenation easier? It seems like it will make life more difficult for our users when they want to write & test their own schemas.
2. This file seems very unapproachable to construct by hand, was this auto generated I'm assuming? A couple questions on the format:
  - why is there a `result` and a `results` field for each entry, even when the entries appear to only use one or the other? It seems more intuitive to just have a `results` field that may only contain a single entry
  - Why do we have to specify the log type for each test when they're all for the same log type in this file? Perhaps related to the following issue
  - Why use `---` to create multiple YAML documents?  Can we not create a single YAML document with multiple tests?
  - The `result` seems very difficult to construct without just running the tool, especially with all the whitespace characters which I'm assuming need to be exactly correct (if they don't need to be exactly correct, can we remove them?). I think it would be much better to let the result be a YAML map, which a user might be able to reasonably construct by hand if they wanted to.
  - For all other specification files, we capitalize the first letter of fields. Can we do that here as well?

I think many of our customers would find this difficult to use. A format like this is more aligned to our other detections as code configurations and I think it would be easier for our customers to make use of:

```
LogType: AWS.CloudTrailDigest
Tests:
  - Name: CloudTrail Digest logs
     Input: "<...>"
     Results:
       -
        awsAccountId: 123456789012,
        digestStartTime: 2020-04-21T12:28:23Z,
        digestEndTime: 2020-04-21T13:28:23Z,
        digestS3Bucket: cloudtrail-test-eu-west-1,
        digestS3Object: AWSLogs/123456789012/CloudTrail-Digest/00421T132823Z.json.gz,
        digestPublicKeyFingerprint: f0249abde0f55218ac45bd3750055109,
        digestSignatureAlgorithm: SHA256withRSA,
        p_any_aws_account_ids: '["123456789012"]',
        p_event_time: 2020-04-21T13:28:23Z,
        p_log_type: AWS.CloudTrailDigest
  - Name: CloudTrail Digest logs first record
     Input: "<...>"
     Results:
       -
        awsAccountId: 123456789012,
        digestStartTime: 2020-04-21T12:28:23Z,
        digestEndTime: 2020-04-21T13:28:23Z,
        digestS3Bucket: cloudtrail-test-eu-west-1,
        digestS3Object: AWSLogs/123456789012/CloudTrail-Digest/00421T132823Z.json.gz,
        digestPublicKeyFingerprint: f0249abde0f55218ac45bd3750055109,
        digestSignatureAlgorithm: SHA256withRSA,
        p_any_aws_account_ids: '["123456789012"]',
        p_event_time: 2020-04-21T13:28:23Z,
        p_log_type: AWS.CloudTrailDigestlogType: AWS.CloudTrailDigest
```

Sorry if I missed a previous PR/confluence doc that outline what this looked like, this is the first time I can recall seeing what these look like.

File: schemas/logs/apache/access_combined.yml
Comment (nhakmiller) [inline schemas/logs/apache/access_combined.yml:8]: I assume these are all auto-generated, is it possible to capitalize the first letter of each field? e.g.
```
Schema: Apache.AccessCombined
Parser:
  Native:
    Name: Apache.AccessCombined
Description: Apache HTTP server access logs using the 'combined' format
ReferenceURL: https://httpd.apache.org/docs/current/logs.html#combined
Version: 0
Fields:
```

This would be aligned with how we format every other CLI configuration file for Panther.

**Review by nhakmiller — COMMENTED** [review]
Schemas seem mostly fine (one small question) and the make file seems fine, but I am really not a fan of the test files.

File: schemas/logs/apache/access_combined.yml
Comment (alxarch) [reply schemas/logs/apache/access_combined.yml:8]: Not really possible now unless we delay this by a lot. We would need to have case-insensitive parsing for the schemas and change the JSONSchema rules. I think this will be left like this until we go to a `version: 1`.

File: schemas/logs/aws/tests/cloud_trail_digest_tests.yml
Comment (alxarch) [reply schemas/logs/aws/tests/cloud_trail_digest_tests.yml:1]: These files are just dumped here from the panther repo.
The format is text-based currently to match what the `logtesting` package expects. 
We have discussed with @georgepsarakis of having `result(s)` be non-text. Haven't found the time to do that yet.
We can see if this can happen when we set up the testing for this repo.

File: schemas/logs/aws/tests/cloud_trail_insight_tests.yml
Comment (alxarch) [reply schemas/logs/aws/tests/cloud_trail_insight_tests.yml:121]: @wey-chiang these are all tests copied over from the panther repo. These test the actual way that Panther parsers treat log inputs. Missing fields for example are here because the parser is not set up to catch this field as a domain.

File: schemas/logs/aws/tests/cloud_trail_tests.yml
Comment (alxarch) [reply schemas/logs/aws/tests/cloud_trail_tests.yml:73]: Any such missing field issues would be best addressed in the `panther` repo. Not really in the scope of this PR to review the parser behavior or the tests themselves. We basically just dumped them here from the `panther` repo so we can feed Panther from this repo's Release page.

File: schemas/logs/aws/cloud_trail.yml
Comment (alxarch) [reply schemas/logs/aws/cloud_trail.yml:75]: I updated the exporting function in the pending PR on `panther` repo. Will dump again and update the schema YAMLs today. Have some more things to tweak in the export code.

File: schemas/logs/apache/access_combined.yml
Comment (kostaspap) [reply schemas/logs/apache/access_combined.yml:8]: @nhakmiller note that this is the current logschema format used in custom logs: https://docs.runpanther.io/data-onboarding/custom-log-types#writing-a-log-schema-for-json-logs

File: schemas/logs/aws/tests/waf_web_acl_tests.yml
Comment (georgepsarakis) [reply schemas/logs/aws/tests/waf_web_acl_tests.yml:76]: > p_any_domain_names for web-123456789.us-west-2.elb.amazonaws.com

Note that it would require special logic to apply this extraction only for [the `Host` header](https://github.com/panther-labs/panther/blob/315721e897cf42bb6eab188a51fb7c0a28023215/internal/log_analysis/log_processor/parsers/awslogs/waf_web_acl.go#L98). I am quite new at the codebase, but my understanding is that we can only define a scanner on the field level, and not conditionally according to another field value.

> p_any_aws_arns for arn:aws:wafv2:us-west-2:123456789EXAMPLE:regional/webacl/panther_web_dev/f96fe2f4-a2f4-4c06-a4bc-afd495bd675f

I see this only for specific log parsers (CloudTrail and VPC), so I guess it also requires some custom logic. I believe the same applies to `p_any_aws_account_ids`.

These (and other custom Panther field notes) seem like good observations though, would you mind summarizing them in an issue at https://github.com/panther-labs/panther instead? Thanks!

File: schemas/logs/aws/tests/waf_web_acl_tests.yml
Comment (georgepsarakis) [reply schemas/logs/aws/tests/waf_web_acl_tests.yml:179]: See https://github.com/panther-labs/panther-analysis/pull/187/files#r569369903

File: schemas/logs/aws/tests/waf_web_acl_tests.yml
Comment (georgepsarakis) [reply schemas/logs/aws/tests/waf_web_acl_tests.yml:312]: See https://github.com/panther-labs/panther-analysis/pull/187/files#r569369903

File: schemas/logs/crowdstrike/aid_master.yml
Comment (georgepsarakis) [inline schemas/logs/crowdstrike/aid_master.yml:85]: There is a small correction pending for this field (https://github.com/panther-labs/panther/pull/2585). What would be the best way to handle similar fixes during the transition?

Comment (alxarch) [reply schemas/logs/crowdstrike/aid_master.yml:85]: The process would be to run `mage build:schemas` and then copy over the `out/schemas` dir to this repo and make a PR with the changes. This should be done *before* releasing the first version of this schema though! Changing the type on a field is a bacwards incompatible change and we cannot do it once a schema has been already released. This will make coordination between the repos a bit tricky moving forward.

Comment (georgepsarakis) [reply schemas/logs/crowdstrike/aid_master.yml:85]: Understood, I think a type change would present difficulties either way due to Athena having an inconsistent table definition for certain files (or are you referring to this problem?).

Ran the build command locally, I think I understand what the process is. I'll wait for this PR to be merged and then open a new one with revised schema. Thanks!

Comment (alxarch) [reply schemas/logs/crowdstrike/aid_master.yml:85]: Make sure you re-run the `mage` command before making that PR. I have pending corrections to the process that exports 'native' log types from `panther`.

File: Makefile
Comment (jacknagz) [inline Makefile:40]: @alxarch I'd prefer if we used the `panther_analysis_tool` for this functionality

**Review by jacknagz — COMMENTED** [review]
Hyped for this! Can you drop a link that explains the YAML schema?

File: Makefile
Comment (alxarch) [reply Makefile:40]: Not really possible right now. We rely on this on `panther` repo as well. I would be quite complex to coordinate. We an revisit this when we take a closer look on the automations. AFAIK pat is Python so it will be tricky to do the automations there since tests and validation all rely on Go code. Bash glues everything nicely.

**Comment by alxarch** [comment]
> Hyped for this! Can you drop a link that explains the YAML schema?

The schema is the same as the one for custom logs. The only addition is the `native` parser config which is a workaround until we can express all our go parsers logic with YAML files (i.e. starlark parsers).

File: Makefile
Comment (jacknagz) [reply Makefile:40]: I'm not following - I'm saying use Python for zipping and hashing instead of bash.

File: schemas/logs/apache/access_combined.yml
Comment (nhakmiller) [reply schemas/logs/apache/access_combined.yml:8]: Can I open an issue in Panther to change that, or are you committed to the way it is now? Important to remember that unlike the panther repo this is a heavily end user facing repo, where we can provide a better user experience we should strive to do so.

File: schemas/logs/aws/tests/cloud_trail_digest_tests.yml
Comment (nhakmiller) [reply schemas/logs/aws/tests/cloud_trail_digest_tests.yml:1]: I see. Is it ok with you then if I open an issue in the panther repo with my suggestions?

**Comment by jacknagz** [comment]
@alxarch I was confused about the `native` parser directive, that's why I asked

**Comment by alxarch** [comment]
> @alxarch I was confused about the `native` parser directive, that's why I asked

It is our escape hatch to allow implementing this early, before our customlogs engine is versatile enough to handle all our native parsers

File: schemas/logs/apache/access_combined.yml
Comment (alxarch) [reply schemas/logs/apache/access_combined.yml:8]: > Can I open an issue in Panther to change that, or are you committed to the way it is now? Important to remember that unlike the panther repo this is a heavily end user facing repo, where we can provide a better user experience we should strive to do so.

It is quite more complex to achieve this because of the YAML specs already stored in DDB. We would also have to update the whole JSONSchema that validates specs and change and re-test much of the customlogs-related pipeline. I would prefer to defer this for when we have time to do all that.

File: Makefile
Comment (alxarch) [reply Makefile:40]: I am already stretched thin having to coordinate between 2 repositories for an under-development feature with many moving parts. We can explore moving this to the pat repository once we have an MVP working.

File: schemas/logs/aws/cloud_trail.yml
Comment (kostaspap) [reply schemas/logs/aws/cloud_trail.yml]: > I updated the exporting function in the pending PR on `panther` repo. Will dump again and update the schema YAMLs today. Have some more things to tweak in the export code.

@alxarch was this done?

File: schemas/logs/aws/tests/cloud_trail_digest_tests.yml
Comment (kostaspap) [reply schemas/logs/aws/tests/cloud_trail_digest_tests.yml:1]: > I see. Is it ok with you then if I open an issue in the panther repo with my suggestions?

That sounds reasonable - @alxarch any concerns?

**Comment by kostaspap** [comment]
@nhakmiller @jacknagz Some of the comments in this PR seem like iterative improvements/changes in the custom log schemas, some of the planned to be addressed in the stories of the epic: https://app.zenhub.com/workspaces/panther-5e4f99c4fbfd0c5803686679/issues/panther-labs/panther/2434

Are there some big issues that you think should be addressed in order to get approval for this PR?

**Comment by nhakmiller** [comment]
> @nhakmiller @jacknagz Some of the comments in this PR seem like iterative improvements/changes in the custom log schemas, some of the planned to be addressed in the stories of the epic: https://app.zenhub.com/workspaces/panther-5e4f99c4fbfd0c5803686679/issues/panther-labs/panther/2434
> 
> Are there some big issues that you think should be addressed in order to get approval for this PR?

Ah my apologies, I had forgot that this repo doesn't allow just any engineer to approve changes, and I was deferring to you to approve. I can approve this, yes the changes discussed appear to be more related to `panther` than to this repo.

**Review by nhakmiller — APPROVED** [review]

**Review by kostaspap — APPROVED** [review]

**Pull request merged**
